### PR TITLE
network: fix UBSAN issue

### LINKED ISSUE #12724: test-networkd-conf seems to fail under UBSan on i386
```sh
$ UBSAN_OPTIONS=print_stacktrace=1:print_summary=1:halt_on_error=1 ./build/test-networkd-conf
...
filename:1: An address '1.2.3.4' is specified without prefix length. The behavior of parsing addresses without prefix length will be changed in the future release. Please specify prefix length explicitly.
../src/network/networkd-address.c:791:81: runtime error: shift exponent 32 is too large for 32-bit type 'long unsigned int'
    #0 0x53e0c9 in config_parse_address ../src/network/networkd-address.c:791
    #1 0x530c56 in test_config_parse_address_one ../src/network/test-networkd-conf.c:178
    #2 0x531591 in test_config_parse_address ../src/network/test-networkd-conf.c:199
    #3 0x531967 in main ../src/network/test-networkd-conf.c:225
    #4 0xb5f0f285 in __libc_start_main (/lib/i386-linux-gnu/libc.so.6+0x18285)
    #5 0x52e7d0  (/home/vagrant/systemd/build/test-networkd-conf+0xa37d0)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../src/network/networkd-address.c:791:81 in
```

### PR DESCRIPTION
**Pull request by yuwata**
Not sure, but hopefully fixes #12724.

cc @evverx.

### CODE REVIEW

**Comment by evverx** [comment]
It fixes the issue in the sense that there is nothing left for UBSan to complain about :-) But I'm not very familiar with that part of the code. Is it OK not to set `n->broadcast.s_addr` is `n->prefixlen < 32`?

**Comment by mrc0mmand** [comment]
And exactly for these kind of issues I'd like to have CI (at least the ASan+UBsan job) on alternative architectures. My original dream was to achieve this using the Fedora Rawhide CI, but I haven't received any *useful response from anyone I asked.

Maybe @dperpeet has any news since the last discussion?

\* useful in regards of nudging this dream a little bit closer to the reality

**Comment by yuwata** [comment]
@evverx I noticed that the value is only used when the prefix length is less than or equal to `30` :-)

File: src/network/networkd-address.c [hunk 1]
```diff
@@ -787,7 +787,7 @@ int config_parse_address(const char *unit,
         else
                 n->in_addr_peer = buffer;
 
-        if (n->family == AF_INET && n->broadcast.s_addr == 0)
+        if (n->family == AF_INET && n->broadcast.s_addr == 0 && n->prefixlen <= 30)
                 n->broadcast.s_addr = n->in_addr.in.s_addr | htonl(0xfffffffflu >> n->prefixlen);
```
Comment (keszybz) [inline src/network/networkd-address.c:791]: This seems to be off-by-one:
```pycon
>>> 0xffffffff >> 31
1
```
Maybe write this as `... && n->prefixlen < 32`.

Comment (yuwata) [reply src/network/networkd-address.c:791]: When `prefixlen == 31` there are no broadcast address. So, it is not necessary to set that here.
You can see that networkd-address.c uses the broadcast address only when `prefixlen <= 30`.

Comment (keszybz) [reply src/network/networkd-address.c:791]: Hmm, when prefixlen==31, then we have two addresses, no? `b...0` and `b...1`. I know this is not very useful, but I think we should allow this for consistency.

Comment (yuwata) [reply src/network/networkd-address.c:791]: It is used for point-to-point. https://tools.ietf.org/html/rfc3021.
Please see also PR #6938 and RFE #4038.

**Comment by evverx** [comment]
@mrc0mmand regarding ASan (and i386 in particular) I think that https://github.com/systemd/systemd/pull/12608 is a much more promising route. I kind of put it off (mainly because I haven't figured out a neat way to turn off libmount and libcap) but feel free to pick up where I left off if you're interested.

**Comment by mrc0mmand** [comment]
> @mrc0mmand regarding ASan (and i386 in particular) I think that #12608 is a much more promising route. I kind of put it off (mainly because I haven't figured out a neat way to turn off libmount and libcap) but feel free to pick up where I left off if you're interested.

That looks indeed promising. By the way, does oss-fuzz/FuzzBuzz runs anything on a non-x86_64 systems or is this primarily aimed for the Ubuntu CI? I guess the linked thread from #12608 answer this question for oss-fuzz, but I'm completely unfamiliar with FuzzBuzz.

File: src/network/networkd-address.c (hunk 1 above)
Comment (keszybz) [reply src/network/networkd-address.c:791]: OK. Thanks for the explanation.

**Pull request merged**